libs/vkd3d/memory.c

```diff
@@ -637,15 +637,112 @@ static void vkd3d_memory_allocator_remove_chunk(struct vkd3d_memory_allocator *a
     vkd3d_memory_chunk_destroy(chunk, device, allocator);
 }
 
+static void vkd3d_memory_allocator_cleanup_clear_queue(struct vkd3d_memory_allocator *allocator, struct d3d12_device *device)
+{
+    struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    unsigned int i;
+
+    for (i = 0; i < VKD3D_MEMORY_CLEAR_COMMAND_BUFFER_COUNT; i++)
+        VK_CALL(vkDestroyCommandPool(device->vk_device, clear_queue->vk_command_pools[i], NULL));
+
+    VK_CALL(vkDestroySemaphore(device->vk_device, clear_queue->vk_semaphore, NULL));
+
+    vkd3d_free(clear_queue->allocations);
+    pthread_mutex_destroy(&clear_queue->mutex);
+}
+
+static HRESULT vkd3d_memory_allocator_init_clear_queue(struct vkd3d_memory_allocator *allocator, struct d3d12_device *device)
+{
+    struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkSemaphoreTypeCreateInfoKHR semaphore_type_info;
+    VkCommandBufferAllocateInfo command_buffer_info;
+    VkCommandPoolCreateInfo command_pool_info;
+    VkSemaphoreCreateInfo semaphore_info;
+    unsigned int i;
+    VkResult vr;
+    HRESULT hr;
+    int rc;
+
+    /* vkd3d_memory_allocator_init will memset the entire
+     * clear_queue struct to zero prior to calling this */
+    clear_queue->next_signal_value = 1;
+
+    if ((rc = pthread_mutex_init(&allocator->mutex, NULL)))
+        return hresult_from_errno(rc);
+
+    command_pool_info.sType = VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO;
+    command_pool_info.pNext = NULL;
+    command_pool_info.flags = VK_COMMAND_POOL_CREATE_TRANSIENT_BIT;
+    command_pool_info.queueFamilyIndex = device->queues[VKD3D_QUEUE_FAMILY_INTERNAL_COMPUTE]->vk_family_index;
+
+    command_buffer_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+    command_buffer_info.pNext = NULL;
+    command_buffer_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+    command_buffer_info.commandBufferCount = 1;
+
+    for (i = 0; i < VKD3D_MEMORY_CLEAR_COMMAND_BUFFER_COUNT; i++)
```

Comment:
There isn't much of a gain to use N command pools here. We can just as easily use 1 pool and just reset the command buffers individually.

Reply:
The queue code which does initial transitions uses that pattern as well fwiw, so might be good for consistency's sake.

Reply:
Fixed.